open-craft/sprints
# [BB-1624] Send ratio/budget alerts and calculate cell budgets

This adds periodic alerts for the budget/sustainability overflows.
It also changes cell budgets to be counted from cell-specific keys instead of cell members, so if somebody from SE logs time on BB ticket, this time is accounted to BB.


## Testing instructions 
1. Set `.env` file (you can find it in a comment on the ticket).
1. Start the backend with `docker-compose -f local.yml up --build`.
1. Create superuser with `docker-compose -f local.yml exec django python manage.py createsuperuser`
1. Run migrations with `docker-compose -f local.yml migrate`.
1. (**Running this will take around 962 seconds**, so you can just leave it running in the background) Go to Django shell: `docker-compose -f local.yml exec django python manage.py shell_plus` and run:
   ```
      from sprints.sustainability.tasks import *
      validate_worklog_cache.delay(long_term=True, force_regenerate=True)
   ```
1. Navigate to frontend and run `npm i && npm start`.
1. Log in.
1. To test the cell budgets:
a) change both `From` and `To` dates to `2019/11/26`,
b) check `Period Spent` of Automation account for [Bebop](http://localhost:3000/board/26) (it should be 3),
c) check that the same as above for [Daniel](http://localhost:3000/board/24/user/Daniel%20Clemente%20Laboreo) is also 3.
1. To test notifications:
a) log into Django admin,
b) add [cell](http://localhost:8000/admin/sustainability/cell/) for `Bebop`,
c) add [budget](http://localhost:8000/admin/sustainability/budget/) for any account that has non-zero `YTD Spent`,
d) go to Django shell: `docker-compose -f local.yml exec django python manage.py shell_plus` and run:
      ```
         from sprints.sustainability.tasks import *
         send_email_alerts()
      ```
   e) check that the emails are outputted.
